MCO-1443: Promote onclusterbuild to GA

yuqi-zhang opened the pull request:
Opened for testing. Based on guidance this is currently the first step: create new v1 API, gate remains off, v1 API is excluded from the image manifests

Also adds in: https://github.com/openshift/api/pull/2089

yuqi-zhang commented:
/test all

yuqi-zhang commented:
/test all

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+type MachineConfigPoolReference struct {
+	// name of the MachineConfigPool object.
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
```
We can now use the CEL format validations for this

yuqi-zhang replied:
Same comment as above, temporarily reverted until we can get it working

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+	// name is the name of the secret used to push or pull this MachineOSConfig object.
+	// this secret must be in the openshift-machine-config-operator namespace.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
```
CEL format validation here too

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+	// +kubebuilder:default:=noarch
+	// +optional
+	ContainerfileArch ContainerfileArch `json:"containerfileArch"`
+	// content is the custom content to be built
```
What is this? Can we expand this documentation?

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+
+type MachineOSImageBuilder struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// +kubebuilder:default:=PodImageBuilder
```
Are we sure we want this defaulted? What if we need to change this in the future as we decide there's a better, default image build method that doesn't rely on today's pod based image builder?

yuqi-zhang replied:
Is it difficult to change the default here in the API if we wanted to change it in the future?

JoelSpeed replied:
When you have a default inside the API, you have to consider that this value will be written out to disk, and so, any existing instance of this API, will get this value and be stuck with it. If you wanted to, say, change blanketly even those instances that are already existing, then this makes it very difficult to change in the future.

We typically have a patten of "when omitted, this means no opinion, and the platform is left to choose a reasonable default, which is subject to change over time", which may be a better fit if you want a default here. 

That said, I think the latest version of the API requires the user to specify this explicitly, which I think makes this a moot point

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+	// +listType=map
+	// +listMapKey=containerfileArch
+	// +kubebuilder:validation:MinItems=0
+	// +kubebuilder:validation:MaxItems=7
```
This should be explained in the godoc

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+	// +kubebuilder:validation:XValidation:rule=`((self.split(':').size() == 2 && self.split(':')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$')) || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$'))`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme ...
+	// +kubebuilder:validation:Required
+	RenderedImagePushspec string `json:"renderedImagePushspec"`
+	// releaseVersion is associated with the base OS Image. This is the version of Openshift that the Base Image is associated with.
```
This line almost says the same thing twice, perhaps this should consolidate into one?

JoelSpeed reviewed (commented):
Can we take this promotion opportunity to go through the API thoroughly and improve validations and godocs please

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,171 @@
+	// version tracks the newest MachineOSBuild for each MachineOSConfig
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Required
+	Version int64 `json:"version"`
```
I think the original intent was that they track a corresponding MachineOSConfig generation to do rebuilds, but in techpreview the implementation hasn't leveraged either field (hard set to 1). Let me check that and get back to you.

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,171 @@
+	// +optional
+	BuilderReference *MachineOSBuilderReference `json:"builderReference"`
+	// relatedObjects is a list of objects that are related to the build process.
+	RelatedObjects []ObjectReference `json:"relatedObjects,omitempty"`
```
Added, is listype also required?

JoelSpeed replied:
Listtype should also be added yes, I would expect this is probably a map type with several map keys?

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,171 @@
+type ObjectReference struct {
+	// group of the referent.
+	// +kubebuilder:validation:Required
+	Group string `json:"group"`
```
I think this whole Objectreference was copied from https://github.com/openshift/api/blob/master/config/v1/types_cluster_operator.go#L96 . I'll try to add some based on insights object

yuqi-zhang commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+	// machineConfigPool is the pool which the build is for
+	// +kubebuilder:validation:Required
+	MachineConfigPool MachineConfigPoolReference `json:"machineConfigPool"`
+	// buildInputs is where user input options for the build live
+	// +kubebuilder:validation:Required
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// buildOutputs is where user input options for the build live
+	// +optional
+	BuildOutputs BuildOutputs `json:"buildOutputs,omitempty"`
```
Let me check on that second point

yuqi-zhang commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// observedGeneration represents the generation observed by the controller.
+	// this field is updated when the user changes the configuration in BuildSettings or the MCP this object is associated with.
```
It looks like we're just incrementing (?) this when a build is completed in reaction to changes to the MOSConfig object, but I'm not sure if that's the right use here?

yuqi-zhang commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+	// must live in the openshift-machine-config-operator namespace if provided.
+	// defaults to using the cluster-wide pull secret if not specified.
+	// +optional
+	BaseImagePullSecret ImageSecretObjectReference `json:"baseImagePullSecret,omitempty"`
```
Removed, but I'm not sure I understand what you mean by the field within. Could you elaborate?

JoelSpeed replied:
Is the `name` field within the `ImageSecretObjectReference` struct considered to be required?

yuqi-zhang replied:
The CEL currently requires the name field. If you provide a reference, we expect you to name the secret to point to.

yuqi-zhang commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+	// host[:port][/namespace]/name:<tag> or svc_name.namespace.svc[:port]/repository/name:<tag>
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=447
+	// +kubebuilder:validation:XValidation:rule=`((self.split(':').size() == 2 && self.split(':')[1].matches('^([a-zA-Z0-9-./:])+$')) || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$'))`,message="the OCI Image reference must end with a valid :<tag>, where '<digest>' is 64 characters long and '<tag>' is any valid string  Or it must be a valid .svc followed by a port, repository ...
```
I don't think this can be a digest. Removed for now, but will check (in conjunction with your point above)

yuqi-zhang commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+	// It will come in the format: 4.16.0-0.nightly-2024-04-03-065948 or any valid release. The MachineOSBuilder populates this field and validates that this is a valid stream.
+	// This is used as a label in the dockerfile that builds the OS image.
+	// +optional
+	ReleaseVersion string `json:"releaseVersion,omitempty"`
```
As in, add validation for the string?

JoelSpeed replied:
Yeah, I'd prefer to see some sort of CEL validation that validates in the input up front

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,171 @@
+type RenderedMachineConfigReference struct {
+	// name is the name of the rendered MachineConfig object.
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
```
I think I misunderstood how this is supposed to be used. Is there an example somewhere?

JoelSpeed replied:
The general format is
```
// +kubebuilder:validation:XValidation:rule="<CEL rule>",messageExpression="<expression>"
```

So something like

```
// +kubebuilder:validation:XValidation:rule="format.dns1123Subdomain().validate(self).hasValue()",messageExpression="format.dns1123Subdomain.validate(self).value().join("\n")"
```

yuqi-zhang replied:
Right, I tried following https://github.com/kubernetes/apiserver/blob/master/pkg/cel/library/format.go but it seems to fail that `format` is not defined? https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_api/2090/pull-ci-openshift-api-master-verify-crd-schema/1856783630089588736

JoelSpeed replied:
Ahh! That might be a tooling issue, the format library is pretty new (I believe shipped in 1.31, will have to double check), and it's possible that the schema checker is not using an up-to-date enough dependency.

Out of interest, did you also try applying the schema to a 4.18 cluster? If it doesn't work there, then it's actually 1.32 that its turned on and we will have to wait to use it

yuqi-zhang replied:
Just to circle back on this comment, I can confirm that 4.18 currently doesn't support this schema. We will most likely still use the old validation method for now.

yuqi-zhang commented:
Updated based on comments:
     - Removed Version and ConfigGeneration from MOSB as they were unused
     - Updated relatedobjects list
     - Changed all optional,omitempty structs to pointers
     - Removed default for ImageBuilderType, but keeping default build arch
       to noarch as we don’t foresee changing that.
     - Fixed RenderedImagePushspec validators to match description

Also temporarily reverted using format.dns1123subdomain() library while figuring out how it works

yuqi-zhang commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,228 @@
+	// this arch is optional. If the user does not specify an architecture, it is assumed
+	// that the content can be applied to all architectures, or in a single arch cluster: the only architecture.
+	// +kubebuilder:validation:Enum:=arm64;amd64;ppc64le;s390x;aarch64;x86_64;noarch
+	// +kubebuilder:default:=noarch
```
We decided that this is fine to default, since we're very unlikely to ever change this

yuqi-zhang commented:
/test e2e-aws-serial-techpreview

umohnani8 commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -234,6 +235,6 @@ type ImageSecretObjectReference struct {
-	// describes that the machine-os-builder will use a custom pod builder that uses buildah
-	PodBuilder MachineOSImageBuilderType = "PodImageBuilder"
+	// describes that the machine-os-builder will use a Job to spin up a custom pod builder that uses buildah
+	PodBuilder MachineOSImageBuilderType = "JobImageBuilder"
```
Should this be renamed to `JobBuilder` as well?

cheesesashimi commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -104,16 +104,16 @@ type MachineOSBuildStatus struct {
 	// +unionMember
 	// +optional
-	PodImageBuilder *ObjectReference `json:"buildPod,omitempty"`
+	JobImageBuilderStatus *ObjectReference `json:"jobImageBuilderStatus,omitempty"`
```
I'm not sure that `BuilderStatus` makes sense here because this is mostly just a reference to the build executor which doesn't include any kind of status message (to my knowledge). Maybe something like `JobImageBuilderRef` or something like that would be more appropriate?

djoshy replied:
I think the comment above this field might need an update to reflect that it is a reference as well. 

Stepping back a bit, I'm guessing the reason we are using a union here is we could have different builders, and they could need different types(other than `ObjectReference`) to represent their reference? If we are always going to use `ObjectReference` for multiple kind of builders, then maybe we don't even need to have to a union discriminator? It could be just two fields within this struct as Zack suggested: type and ref.

cheesesashimi replied:
Those values are arguably common to all image builders. There could be some additional information that may be desired specific to each type of image builder, but we can add that onto this later as a specific type.

cheesesashimi replied:
And taking a step back from this, if the `ObjectReference` includes GVK (Group Version Kind) information, we could potentially eliminate the `ImageBuilderType` field here since I see that as being primarily user-facing.

yuqi-zhang replied:
Ok, so instead of having individual ImageBuilderRef's, I'll just make it one optional field here to encompass all potential object references.

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+// ObjectReference contains enough information to let you inspect or modify the referred object.
+type ObjectReference struct {
```
We could add one, the corev1 default does have it, but some of the locally defined types do not. Do you see a need for the UID in the status? I thought this was mostly so a user can cross-reference the object from here, and if they need more information, they can look directly at the referenced object (Job)

cheesesashimi replied:
After some offline discussion about the use-case I had in mind, I don't think we need the UID after all.

cheesesashimi commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,240 @@
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@').size() == 2 && self.split('@')[1].matches('^sha256:[a-f0-9]{64}$'))`,message="the OCI Image reference must end with a valid '@sha256:<digest>' suffix, where '<digest>' is 64 characters long"
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$'))`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme"
+	// +optional
+	CurrentImagePullspec string `json:"currentImagePullspec,omitempty"`
```
I feel like it might be useful to have a reference to the currently active MachineOSBuild here. In practice, I've been using an annotation for this purpose, but I think having it on the status would be better.

yuqi-zhang replied:
Added :+1:

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// machineOSImageBuilder describes which image builder will be used in each build triggered by this MachineOSConfig
+	// Currently supported type(s): JobImageBuilder
+	// +kubebuilder:validation:Required
+	ImageBuilder *MachineOSImageBuilder `json:"imageBuilder"`
```
Should not be a pointer for a required struct

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// +kubebuilder:validation:Required
+	ImageBuilder *MachineOSImageBuilder `json:"imageBuilder"`
+	// renderedImagePushSecret is the secret used to connect to a user registry.
+	// the final image push and pull secrets should be separate for security concerns. If the final image push secret is somehow exfiltrated,
+	// that gives someone the power to push images to the image repository. By comparison, if the final image pull secret gets exfiltrated,
+	// that only gives someone to pull images from the image repository. It's basically the principle of least permissions.
+	// this push secret will be used only by the MachineConfigController pod to push the image to the final destination. Not all nodes will need to push this image, most of them
+	// will only need to pull the image in order to use it.
```
I wonder if we can make this more direct, but still convey the same message, what about?
```suggestion
	// The final image push and pull secrets should be separate and assume the principal of least privilege.
	// The push secret with write privilege is only required to be present on the node hosting the MachineConfigController pod.
	// The pull secret with read only privileges is required on all nodes.
	// By separating the two secrets, the risk of write credentials becoming compromised is reduced.
```

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	RenderedImagePushSecret ImageSecretObjectReference `json:"renderedImagePushSecret"`
+	// renderedImagePushspec describes the location of the final image.
+	// the MachineOSConfig object will use the in cluster image registry configuration.
+	// if you wish to use a mirror or any other settings specific to registries.conf, please specify those in the cluster wide registries.conf.
```
How would I do this?

yuqi-zhang replied:
updated

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// +kubebuilder:validation:XValidation:rule=`self.matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?(/[a-zA-Z0-9-_]{1,61})*/[a-zA-Z0-9-_.]+:[a-zA-Z0-9._-]+$') || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$')`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme. Or it must be a valid .svc follo ...
+	// +kubebuilder:validation:Required
+	RenderedImagePushspec string `json:"renderedImagePushspec"`
+	// releaseVersion is an Openshift release version which the base OS image is associated with.
```
Does this need to be in the spec? Sounds like we aren't expecting users to populate this?

yuqi-zhang replied:
Will check if we will allow overrides

cheesesashimi replied:
Right now, this field gets populated with the value from the `RenderedImagePushspec` on the MachineOSConfig with the name of the MachineOSBuild in place of the tag. In other words, `hostname.com/org/repo:tag` becomes `hostname.com/org/repo:<MachineOSBuild Name>`. I'm not sure if we should allow overrides or not, so if we want to remove this for now, I'm fine with that.

yuqi-zhang replied:
Hmm, I'm not sure I understand. I interpreted the original question as whether `releaseVersion` is a required field. The godoc currently is taking about the Openshift release, and the MCO code adds it to the containerfile as a `releaseversion` label. Do you think that is still a pattern we want to do?

JoelSpeed replied:
Is MachineOSConfig something a user would create? What happens if they populate `releaseVersion` with something, potentially nonsense that doesn't make sense on the cluster? Is the MAO going to overwrite it with the correct value? Are there multiple correct values?

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// It will come in the format: 4.16.0-0.nightly-2024-04-03-065948 or any valid release. The MachineOSBuilder populates this field and validates that this is a valid stream.
+	// This is used as a label in the dockerfile that builds the OS image.
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:Pattern=`^([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9])(\.([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\-]{0,61}[a-zA-Z0-9]))*$`
```
What's the pattern here? Is this one of the standard formats like DNS1123Subdomain?

yuqi-zhang replied:
Yes, will update this validation too

yuqi-zhang replied:
It looks like this is just how all "name" type objects are validated across the MCO namespaced objects. So probably some precedent there.

If I'm grepping correctly, the other place I see this is `Host` field under route, which is talking about `DNS952` instead. Maybe all the MCO validation was inaccurate?

JoelSpeed replied:
Are there any tests associated to the validation that show some examples of things that do and do not pass?

Breaking down the regex, it looks like it's the same thing, except it's limiting each segment to 63 chars, where I don't think 1123 does do that 🤔 So, moving to DNS1123 would be a relaxation of the validation, but would not be incompatible I don't think

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	ReleaseVersion string `json:"releaseVersion,omitempty"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// this is also commonly called a Dockerfile and you can treat it as such. The content is the content of your Dockerfile.
+	// you can specify up to 7 containerFiles
```
How did we conclude on 7? Is there a specific reason for this limit?

yuqi-zhang replied:
Somewhat arbitrary cap. Most users should be only providing one, but we would like to give them the flexibility to add a few more

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// +optional
+	ReleaseVersion string `json:"releaseVersion,omitempty"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// this is also commonly called a Dockerfile and you can treat it as such. The content is the content of your Dockerfile.
```
Perhaps we want to describe the content of the file, is there actually a spec for Dockerfile/ContainerFile that is named/that we could reference?

yuqi-zhang replied:
Hmm, would linking to e.g. https://github.com/containers/common/blob/main/docs/Containerfile.5.md work?

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+}
+
+// BuildOutputs holds all information needed to handle booting the image after a build
+// +union
```
This doesn't look like a union type?

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// imageBuilderType specifies the backend to be used to build the image.
+	// +kubebuilder:validation:Enum:=JobImageBuilder
+	// Valid options are: JobImageBuilder
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
```
```suggestion
	// +required
	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
```

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+
+// MachineOSContainerfile contains all custom content the user wants built into the image
+type MachineOSContainerfile struct {
+	// containerfileArch describes the architecture this containerfile is to be built for
+	// this arch is optional. If the user does not specify an architecture, it is assumed
```
Punctuation is important to ensure `oc explain` output makes sense
```suggestion
	// containerfileArch describes the architecture this containerfile is to be built for.
	// This arch is optional. If the user does not specify an architecture, it is assumed
```

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// containerfileArch describes the architecture this containerfile is to be built for
+	// this arch is optional. If the user does not specify an architecture, it is assumed
+	// that the content can be applied to all architectures, or in a single arch cluster: the only architecture.
+	// +kubebuilder:validation:Enum:=arm64;amd64;ppc64le;s390x;aarch64;x86_64;noarch
```
These should be PascalCase normally, and not lowercased like this, the proper name for arm64 is ARM64 anyway, for example

Also, AMD64 and x86_64 are the same thing? AArch64 and ARM64 are the same?

Is noarch a special constant for this implementation or something borrowed from elsewhere?

yuqi-zhang replied:
Will check on this point

cheesesashimi replied:
> AMD64 and x86_64 are the same thing? AArch64 and ARM64 are the same?

Yes they are. For some reason, `$ uname -m` returns `x86_64`, but anything Golang-related wants `amd64`. The same applies to AArch64 and ARM64.

> Is noarch a special constant for this implementation?

It is. Basically, if a Containerfile is associated with noarch, it means that the same Containerfile should be able to be built for any supported architecture. There are situations where one may want to supply a different Containerfile for a different architecture. The most common example I can think of off the top of my head is when someone is installing an architecture-specific script or tool and doing something like `curl "https://host.com/$(uname -m)/tool.rpm"` doesn't cut it.

JoelSpeed replied:
> Yes they are. For some reason, $ uname -m returns x86_64, but anything Golang-related wants amd64. The same applies to AArch64 and ARM64.

We should make sure we only have one version for each real architecture (the go one typically), to avoid duplication within the enum. If you need to convert, convert within the operator.

> Basically, if a Containerfile is associated with noarch, it means that the same Containerfile should be able to be built for any supported architecture.

When there are many containerfiles configured, are they supposed to be different architectures? What happens if I were to specify multiple noarch containerfiles in the list?

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	//   RUN rpm-ostree install tree && \
+	//     ostree container commit
+	// +kubebuilder:validation:Required
+	Content string `json:"content"`
```
We should add a length limit to this string, and avoid arbitrarily long content.

Is 4kb enough? 

```suggestion
	// +kubebuilder:validation:MaxLength=4096
	Content string `json:"content"`
```

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	Content string `json:"content"`
+}
+
+type ContainerfileArch string
```
For the purposes of openapi generation
```suggestion
// +enum
type ContainerfileArch string
```
Should we move the kubebuilder enum tag here too?

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// +kubebuilder:default:=noarch
+	// +optional
+	ContainerfileArch ContainerfileArch `json:"containerfileArch"`
+	// content is an embedded Containerfile/Dockerfile that defines the contents to be built into your image.
```
Is there a spec name for the format within a Containerfile/Dockerfile?

yuqi-zhang replied:
Maybe just the containerfile spec? (See above link)

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	Name string `json:"name"`
+}
+
+type MachineOSImageBuilderType string
```
Add the `+enum` and perhaps move the kubeubilder enum tag here as well

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// +kubebuilder:validation:MaxLength=447
+	// +kubebuilder:validation:XValidation:rule=`self.matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?(/[a-zA-Z0-9-_]{1,61})*/[a-zA-Z0-9-_.]+:[a-zA-Z0-9._-]+$') || self.matches('^[^.]+\\.[^.]+\\.svc:\\d+\\/[^\\/]+\\/[^\\/]+:[^\\/]+$')`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme. Or it must be a valid .svc follo ...
+	// +kubebuilder:validation:Required
+	RenderedImagePushspec string `json:"renderedImagePushspec"`
```
Is a `Pushspec` a thing or should this be `PushSpec`?

yuqi-zhang commented on hack/update-payload-crds.sh:
```diff
@@ -21,7 +21,11 @@ crd_globs="\
     config/v1/zz_generated.crd-manifests/0000_10_openshift-controller-manager_01_builds*.crd.yaml
     operator/v1/zz_generated.crd-manifests/0000_50_openshift-controller-manager_02_openshiftcontrollermanagers*.crd.yaml
-    machineconfiguration/v1/zz_generated.crd-manifests/*.crd.yaml
+    machineconfiguration/v1/zz_generated.crd-manifests/0000_80_machine-config_01_containerruntimeconfigs*.crd.yaml
+    machineconfiguration/v1/zz_generated.crd-manifests/0000_80_machine-config_01_controllerconfigs*.crd.yaml
+    machineconfiguration/v1/zz_generated.crd-manifests/0000_80_machine-config_01_kubeletconfigs*.crd.yaml
+    machineconfiguration/v1/zz_generated.crd-manifests/0000_80_machine-config_01_machineconfigpools*.crd.yaml
+    machineconfiguration/v1/zz_generated.crd-manifests/0000_80_machine-config_01_machineconfigs*.crd.yaml
```
They will be once we switch the MCO code to use the new CRDs. Right now, we're just generating the new API here, and adapting the MCO code to use it with the new vendor. Then we will open a PR here to graduate the FG + have new new v1 CRDs generate, and test with both PRs.

The concern is that if I do generate with the new v1 CRDs off the bat, all the techpreview jobs would fail since the MCO is still using the old API/CRD. I could be mis-understanding that though! In that case, we can switch directly over (I made https://github.com/openshift/api/pull/2100/files to do that)

JoelSpeed replied:
Ahh right, yes, good point! In that case you're right, you'll need to do this in a later step. Can you add a TODO or something in here so we remember that this is why we are doing this please

yuqi-zhang commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -104,9 +104,10 @@ type BuildInputs struct {
-	// +kubebuilder:validation:Required
-	BaseImagePullSecret ImageSecretObjectReference `json:"baseImagePullSecret"`
+	// must live in the openshift-machine-config-operator namespace if provided.
+	// defaults to using the cluster-wide pull secret if not specified.
```
The cluster pull secret is the one in `openshift-config` namespace (just a secret called pull-secret). There shouldn't be any security implications - this is widely used, and is on-disk for every node as well as referenced in various MCO objects, and doesn't give any push access in most cases.

JoelSpeed replied:
So maybe this should read something along the lines of, though, I'm not sure what registries it actually has read access to, does it have access to the in-cluster registry?
```suggestion
	// Defaults to using the cluster-wide pull secret, that was provided at install time.
	// The default secret will provide read only access to registries <such as in cluster? RH owned? I don't know what this actually gives>.
```

yuqi-zhang replied:
I'll update the description in the v1 API. It does not, but we expect users to use the base shipped payload RHCOS image as the base image most of the time, so it should not be necessary.

yuqi-zhang commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -104,9 +104,10 @@ type BuildInputs struct {
+	// must live in the openshift-machine-config-operator namespace if provided.
+	// defaults to using the cluster-wide pull secret if not specified.
+	// +optional
+	BaseImagePullSecret ImageSecretObjectReference `json:"baseImagePullSecret,omitempty"`
```
I think I updated that by mistake. Will remove. It should be correct in the v1 API.

Basically the v1 API is updated here, and we will remove the v1alpha1 API after we switch all references, hence why I haven't updated the v1alpha1 API. Is that generally the expected pattern?

JoelSpeed replied:
Yeah generally you can leave the v1alpha1 API alone, I just checked it as you had made changes in it. You shouldn't need to add this here if it's being added strictly to the new v1 API

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,178 @@
+type MachineOSBuildStatus struct {
+	// conditions are state related conditions for the build. Valid types are:
+	// Prepared, Building, Failed, Interrupted, and Succeeded
+	// once a Build is marked as Failed, no future conditions can be set. This is enforced by the MCO.
```
The expectation is that only the MCO will be creating MachineOSBuild objects (so the user cannot directly create a build like that, they must create a config from which build objects are created by the MCO as it does the builds). In that sense, this status is monitored and updated by the MCO. 

Do you think it's strictly necessary given the context?

JoelSpeed replied:
Yeah I would still add the validation. Users have access to APIs, and may do unexpected things. Controllers also have bugs, that make them do weird things. The more we can put into the API validation to make sure it meets our expectations, the better, even in status

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,178 @@
+	// BuilderReference describes which ImageBuilder backend to use for this build
+	// +optional
+	BuilderReference *MachineOSBuilderReference `json:"builderReference"`
+	// relatedObjects is a list of objects that are related to the build process.
```
I think the original intent is that this would be used to track configmap/secrets/etc. objects related to the build. Mostly for reference only

JoelSpeed replied:
Can you expand the godoc a little to explain what users might see in here please, and why

cheesesashimi replied:
Here's an idea for the godoc:

> RelatedObjects is a list of references to ephemeral objects such as ConfigMaps or Secrets that are meant to be consumed while the build process runs. After a successful build or when this MachineOSBuild is deleted, these ephemeral objects should be deleted. However, in the event of a failed build, the objects will not be deleted to allow for inspection and debugging of the failed build process.

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,178 @@
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="buildStart is immutable once set"
+	// +kubebuilder:validation:Required
+	BuildStart metav1.Time `json:"buildStart"`
+	// buildEnd describes when the build ended.
```
The build has not yet completed. The end time is added when the build completes, fails, or is interrupted.

JoelSpeed replied:
```suggestion
	// buildEnd describes when the build ended.
	// When omitted the build is still in progress.
	// It will be populated once the build completes, fails or is interrupted.
```

JoelSpeed replied:
Is there a dependency between start time and end time? We could validate that buildEnd is only added after buildStart, and we could even validate that buildEnd is after buildStart 🤔

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,178 @@
+	// ImageBuilderType describes the image builder set in the MachineOSConfig
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+
+	// ImageBuilderRef is a reference to the object that is managing the image build
```
a kubernetes job object, in this case.

JoelSpeed replied:
So perhaps a `For example, when imageBuilderType is foo, this would point to the bar that does the thing` kind of line would be helpful in this godoc

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,178 @@
+	// Example: "", "apps", "build.openshift.io", etc.
+	// +kubebuilder:validation:Pattern:="^$|^[a-z0-9]([-a-z0-9]*[a-z0-9])?(.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
+	// +kubebuilder:validation:Required
+	Group string `json:"group"`
```
Not sure, set one just in case

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,178 @@
+	// group of the referent.
+	// This value should consist of only lowercase alphanumeric characters, hyphens and periods.
+	// Example: "", "apps", "build.openshift.io", etc.
+	// +kubebuilder:validation:Pattern:="^$|^[a-z0-9]([-a-z0-9]*[a-z0-9])?(.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*$"
```
Ack, will update these

yuqi-zhang commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// buildInputs is where user input options for the build live
+	// +kubebuilder:validation:Required
+	BuildInputs BuildInputs `json:"buildInputs"`
+	// buildOutputs holds all information needed to handle booting the image after a build
+	// +optional
+	BuildOutputs *BuildOutputs `json:"buildOutputs,omitempty"`
```
This is user input. Currently the only field here is a pull secret that the user can optionally provide for the registry to which the image is pushed to, if that's different from the push secret, and different from the cluster pull secret.

yuqi-zhang commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+
+// MachineOSConfigStatus describes the status this config object and relates it to the builds associated with this MachineOSConfig
+type MachineOSConfigStatus struct {
+	// conditions are state related conditions for the config.
```
I'll set this to 5 for now. Doesn't seem like we're leveraging the field in the code yet.

yuqi-zhang commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// +listMapKey=type
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// observedGeneration represents the generation observed by the Machine Config Operator's build controller.
```
will add, this is the last observed generation of the machineosconfig from the MCController side

yuqi-zhang commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=447
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@').size() == 2 && self.split('@')[1].matches('^sha256:[a-f0-9]{64}$'))`,message="the OCI Image reference must end with a valid '@sha256:<digest>' suffix, where '<digest>' is 64 characters long"
+	// +kubebuilder:validation:XValidation:rule=`(self.split('@')[0].matches('^([a-zA-Z0-9-]+\\.)+[a-zA-Z0-9-]+(:[0-9]{2,5})?/([a-zA-Z0-9-_]{0,61}/)?[a-zA-Z0-9-_.]*?$'))`,message="the OCI Image name should follow the host[:port][/namespace]/name format, resembling a valid URL without the scheme"
```
I can try that. Curious how hard it is to switch between different validation types after GA? I thought validation was always safe to switch (unless we would reject something previously accepted?)

JoelSpeed replied:
You can switch, but you have to be certain that switching does not, as you mention, reject something previously accepted. And as such, I'm keen to get things as good as we can get them, using modern tooling, before we merge. Promising to come back and fix them later always has risk

yuqi-zhang commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,243 @@
+	// +optional
+	BaseOSImagePullspec string `json:"baseOSImagePullspec,omitempty"`
+	// baseImagePullSecret is the secret used to pull the base image.
+	// must live in the openshift-machine-config-operator namespace if provided.
+	// defaults to using the cluster-wide pull secret if not specified.
```
As discussed above, I think this is relatively non-ambiguous and is relatively safe. Users would have provided this during install time to get the cluster going so it should be valid

yuqi-zhang commented:
Thanks for the thorough review. Addressed most of the comments, left a few questions, and will follow up on the 3 points above.

I see the o/k PR for adding the format library has merged. Will try to incorporate that into the next set of changes as well.

yuqi-zhang commented:
Pushed another set of updates:
 - Failed and Interrupted will now cause MOSBuild conditions to be immutable
 - Updated Arch enum to be PascalCase (kept x86_64 though)
 - Updated relatedObject go doc based on suggestion
 - Add validation for buildEnd > buildStart (did not add ordering requirement, as buildStart is a required field, and will be applied before or at the same time of buildEnd)
 - Removed conditions field from MOSConfig -> the build object is supposed to reflect conditions instead, so this is not needed at this time
 - Use dns1123 format check for all strings that match, and otherwise switch pattern checks to validation rules where appropriate
 - Updated godocs a bit more for formatting

Tested on the latest 4.18 nightly and can confirm both CRDs are valid with the new format library

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,247 @@
+// MachineOSConfigStatus describes the status this config object and relates it to the builds associated with this MachineOSConfig
+type MachineOSConfigStatus struct {
+	// observedGeneration represents the generation of the MachineOSConfig object observed by the Machine Config Operator's build controller.
+	// +kubebuilder:validation:XValidation:rule="self >= oldSelf || (self == 0 && oldSelf > 0)", message="observedGeneration must not move backwards except to zero"
```
When does it move to zero? 

Also, I think logically `&& oldSelf > 0` here can never be reached, `self >= oldSelf` would already have returned true

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,247 @@
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Required
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// currentImagePullspec is the fully qualified image pull spec used by the MCO to pull down the new OSImage. This must include sha256.
```
```suggestion
	// currentImagePullspec is the fully qualified image pull spec used by the MCO to pull down the new OSImage. This includes the sha256 image digest.
```

JoelSpeed replied:
When is this field populated? After the build? We should mention when this value is updated/changes

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,247 @@
+
+// MachineOSConfigStatus describes the status this config object and relates it to the builds associated with this MachineOSConfig
+type MachineOSConfigStatus struct {
+	// observedGeneration represents the generation of the MachineOSConfig object observed by the Machine Config Operator's build controller.
```
If the generation changes, what does that mean? The MachineOSConfig spec has changed, so the controller needs to trigger a new build?

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,247 @@
+}
+
+// MachineOSConfigStatus describes the status this config object and relates it to the builds associated with this MachineOSConfig
+type MachineOSConfigStatus struct {
```
I'm surprised not to see conditions in this status?

yuqi-zhang replied:
The alpha API had conditions but we removed them here since the status would be the same as on the build object. Do you think there's value in having the same conditions on both?

JoelSpeed replied:
Generally I'd expect to see conditions on most objects, and they would aggregate lower level objects.

For example, what would you do to report and issue in triggering the build? If there's no build object, how does the user know what went wrong

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,247 @@
+}
+
+// BuildInputs holds all of the information needed to trigger a build
+type BuildInputs struct {
```
As a user of this API, how would I know what a BaseOSExtensions vs BaseOS vs Base image means?

I'm wondering if we can clean this API up a bit, there's a lot of, what appears to a naive eye at least, redundancy?
```
images:
  base: <>
  baseOS: <>
  baseOSExtensions: <>
pullSecret:
  name: <>
pushSecret:
  name: <>
builder:
  type: Job
  job:
    name: <>
pushSpec: <>
releaseVersion: <>
containerFiles: []
```
Or if we moved up to spec, some of this might look like
```
sourceImages:
  pullSecret: <> # For pulling these images
  base: <>
  baseOS: <>
  baseOSExtensions: <>
buildOutputs:
  pullSecret: 
    name: <> # For pulling the pushed image
  pushSecret:
    name: <> # For pushing the built image
  imagePushSpec: <>
builder:
  type: Job
  job:
    name: <>
releaseVersion: <>
containerFiles: []
```

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,247 @@
+	// containerfileArch describes the architecture this containerfile is to be built for.
+	// This arch is optional. If the user does not specify an architecture, it is assumed
+	// that the content can be applied to all architectures, or in a single arch cluster: the only architecture.
+	// +kubebuilder:validation:Enum:=ARM64;AMD64;PPC64LE;S390X;AArch64;x86_64;NoArch
```
Why do we have these duplicates? That means an end user can configure the same meaning, in multiple ways, that's confusing, we should avoid that
```suggestion
	// +kubebuilder:validation:Enum:=ARM64;AMD64;PPC64LE;S390X;NoArch
```

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,247 @@
+	// containerFile describes the custom data the user has specified to build into the image.
+	// This is also commonly called a Dockerfile and you can treat it as such. The content is the content of your Dockerfile.
+	// See https://github.com/containers/common/blob/main/docs/Containerfile.5.md for the spec reference.
+	// you can specify up to 7 containerFiles
```
What does it mean to specify multiple images? Why would an end user want to do that?

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,247 @@
+	CurrentImagePullSecret *ImageSecretObjectReference `json:"currentImagePullSecret,omitempty"`
+}
+
+type MachineOSImageBuilder struct {
```
When this struct is used, it's called something like `imageBuilder`, maybe we avoid the repetition here and go for just `type` for the field within here?

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,247 @@
+// the MachineOSBuilder pod validates that the user has provided a valid pool
+type MachineConfigPoolReference struct {
+	// name of the MachineConfigPool object.
+	// Must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character.
```
Should update this to match the others I suggested, including the at most 253 characters part

yuqi-zhang commented on machineconfiguration/v1/tests/machineosconfigs.machineconfiguration.openshift.io/OnClusterBuild.yaml:
```diff
@@ -0,0 +1,178 @@
+          baseImagePullSecret:
+            name: foo
+          baseOSExtensionsImagePullspec: example.io/my-project/image-v1.0_23@sha256:2c3ea52ac3a41c6d58e85977c3149413e3fa4b70eb2397426456863adbf43306
+          renderedImagePushSecret:
+            name: foo
+          renderedImagePushspec: quay.io/mco/renderedImg:latest
```
The user should be specifying where to push to here, so it is user input

JoelSpeed replied:
I agree it's a user input, that's why it's in spec. But it relates to the output of the build, not the input to the build, right?

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,198 @@
+	// BuilderReference describes which ImageBuilder backend to use for this build
+	// +optional
+	BuilderReference *MachineOSBuilderReference `json:"builderReference"`
+	// relatedObjects is a list of references to ephemeral objects such as ConfigMaps or Secrets that are meant to be consumed while the build process runs.
```
The MCO will populate this as it creates the necessary objects. It's a for reference thing only, and no API is further consuming this field, hence in status.

JoelSpeed replied:
Right, I think we can probably update the description a little here. We should explain that the objects are created during the build process/to aid the build process.

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,198 @@
+
+// MachineOSBuilderReference describes which ImageBuilder backend to use for this build
+type MachineOSBuilderReference struct {
+	// ImageBuilderType describes the image builder set in the MachineOSConfig, which in turn describes the builder that the cluster will attempt the build with.
+	// Currently only JobImageBuilder is supported, which will spin up a custom pod builder that uses buildah to build the specified image.
+	// +kubebuilder:validation:Required
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
```
Updated description, but retained the name to better match the MOSC object.

JoelSpeed replied:
I believe I suggested simplfying the names on MOSC as well, why not update in both and remove the redundancy of typing ImageBuilder so many times?

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,198 @@
+	// +kubebuilder:validation:Required
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+
+	// ImageBuilderRef is a reference to the object that is managing the image build
+	// For example, if the imageBuilderType is JobImageBuilder, this will be a reference to the Job object managing the build
+	// +optional
+	ImageBuilderRef *ObjectReference `json:"ImageBuilderRef,omitempty"`
```
This was actually originally an union, but we decided to make it just the normal reference for now, and add fields if needed.

JoelSpeed replied:
Do we have any inspiration for what future builders might look like? Are there any where the ObjectReference wouldn't work?

Bear in mind that adding new fields to ObjectReference isn't so simple, it's used in other places in the API as well, and so if you wanted to add fields within ImageBuidlerRef, you'd then have to change the type here so as not to pollute other areas of the API

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,198 @@
+	// ImageBuilderRef is a reference to the object that is managing the image build
+	// For example, if the imageBuilderType is JobImageBuilder, this will be a reference to the Job object managing the build
+	// +optional
+	ImageBuilderRef *ObjectReference `json:"ImageBuilderRef,omitempty"`
```
While I happen to be looking at this field, API conventions say you shouldn't use `Ref`, and therefore just ImageBuilder. Also, your json tag is not camelCase, it's PascalCase, that needs to be fixed

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,198 @@
+	BuilderReference *MachineOSBuilderReference `json:"builderReference"`
+	// relatedObjects is a list of references to ephemeral objects such as ConfigMaps or Secrets that are meant to be consumed while the build process runs.
+	// After a successful build or when this MachineOSBuild is deleted, these ephemeral objects should be deleted.
+	// However, in the event of a failed build, the objects will not be deleted to allow for inspection and debugging of the failed build process.
```
Will they be deleted when the MachineOSBuild is deleted? Should they always just be deleted when the MachineOSBuild is deleted?

yuqi-zhang commented:
Further API updates
 - Removed BuildOutputs and the CurrentImagePullSecret field (not really needed at this time, we’d prefer if the user would put any additional pull secrets into the cluster-wide object)
 - Removed BuildInputs and lifted all fields to spec, removing:
    - ReleaseVersion
    - BaseOSImagePullSpec
    - BaseOSExtensionsImagePullSpec
   Which will be populated directly from the MCO instead. We will consider re-adding those if there is a need for on cluster pre-builds off new release images in the future.
 - Renamed finalImagePushSpec to digestedImagePushSpec
 - Switched MachineOSBuilderReference back to an union, and renamed the job object unionmember to just “job”
 - Changed “desiredConfig” to “MachineConfig” for clarity
 - Kept ObservedGeneration, but updated the validation, and will fix on the MCO side
 - Removed duplicate arch types, and updated containerfiles to allow at most 1 per arch (minus noarch)

One followup question, since the job builder is an union now, should it just be a job object instead of a objectreference here: https://github.com/openshift/api/pull/2090/files#diff-63de8e8386ec97c788eb321a3c185970ca9751290b40738584bf874c5e6abf4fR128

JoelSpeed commented on machineconfiguration/v1/tests/machineosbuilds.machineconfiguration.openshift.io/OnClusterBuild.yaml:
```diff
@@ -0,0 +1,212 @@
+              reason: "Reason"
+              message: "Message"
+              lastTransitionTime: "2024-11-28T10:00:00Z"
+      expectedError: "status.conditions[0].status: Invalid value: \"string\": once a Failed condition is set, conditions are immutable"
```
```suggestion
      expectedStatusError: "status.conditions[0].status: Invalid value: \"string\": once a Failed condition is set, conditions are immutable"
```

enxebre commented:
can we please link the enhancement that backs this API?

enxebre commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,208 @@
+	// buildEnd is the timestamp corresponding to completion of the builder backend.
+	// When omitted the build has either not been started, or is in progress.
+	// It will be populated once the build completes, fails or is interrupted.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="buildEnd is immutable once set"
```
To honour the description, don't you need immutability to be relaxed here to allow transitioning from unset? e.g. oldSelf == null || self == oldSelf

yuqi-zhang replied:
This works fine generally for all cases. I don't think the validation kicks in until `self` is not null. I've added a test case just in case but it does work.

JoelSpeed replied:
Yep, Jerry is correct, since this is a transition rule (contains oldSelf), it isn't effected until there is an `oldSelf`.

In the future, we may use `optionalOldSelf` to create ratcheting rules, that would change things, but for now, safe assumption

enxebre commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,208 @@
+	// +optional
+	BuildEnd *metav1.Time `json:"buildEnd,omitempty"`
+	// digestedImagePushSpec describes the fully qualified push spec produced by this build.
+	// The format of the push spec is: host[:port][/namespace]/name@sha256:<digest>,
```
can we include regex validation for this?

yuqi-zhang replied:
It is validated. Through the type: https://github.com/openshift/api/pull/2090/files#diff-9ce2b237d44d6d21fe5f87bcdb1e8306601d463ab800c43e002adca3a7a35944R192-R199

yuqi-zhang commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,208 @@
+package v1
+
```
Based on the discussions in the HCP office hours, this will work in the same way that MachineConfigs would through configmap-mapped objects.

The main reasoning being that we would like to decouple management cluster version and the API, such that e.g. a 4.17 management cluster can configure it for a 4.19 hosted cluster, despite the API not being available for 4.17

JoelSpeed replied:
Have you considered an alternative abstraction at the HCP layer that would be namespace scoped? It would look very similar to this API, but expose it in a way that makes sense for HCP.

Is there any intention that, in HCP, these builds exist within the management plane?

enxebre replied:
>The main reasoning being that we would like to decouple management cluster version and the API, such that e.g. a 4.17 management cluster can configure it for a 4.19 hosted cluster, despite the API not being available for 4.17

I think setting a req on the management cluster version for certain features is valid. Besides the API how will the mco controllers watch and react to those configmaps? What Joel suggest above is one possible starting point to answer these questions. Let's make sure we iterate a bit more on this before setting an API in stone.

yuqi-zhang replied:
>Besides the API how will the mco controllers watch and react to those configmaps?

yuqi-zhang commented:
The most direct enhancement is https://github.com/openshift/enhancements/pull/1515. Or are you asking to link it to the code directly?

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,208 @@
+
+	// status describes the last observed state of this machine os build
+	// +optional
+	Status *MachineOSBuildStatus `json:"status"`
```
You need omitempty if this is a pointer.
```suggestion
	Status *MachineOSBuildStatus `json:"status,omitempty"`
```

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,208 @@
+type MachineOSBuildStatus struct {
+	// conditions are state related conditions for the build. Valid types are:
+	// Prepared, Building, Failed, Interrupted, and Succeeded.
+	// Once a Build is marked as Failed or Interrupted, no future conditions can be set.
```
So, question on the mechanics of this, does the controller just see these states and bail out early? There's no point where it's trying to update the Failed or Interrupted conditions even after they've been added?

yuqi-zhang replied:
The controller is the only updator, and prevents itself from resetting the status once in one of the terminal states:

https://github.com/openshift/machine-config-operator/blob/67789532e4569ece95eb98fed1dde03409f065e9/pkg/controller/build/helpers.go#L162

I'll also add "success" to the validations here since you shouldn't be able to transition out of it.

cc @cheesesashimi in case I'm missing something

cheesesashimi replied:
That's correct. Once the MachineOSBuild reaches a terminal state (e.g., success, failed, interrupted), it won't try to do any other updates on it.

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,208 @@
+	// +optional
+	Builder *MachineOSBuilderReference `json:"builder"`
+	// relatedObjects is a list of references to ephemeral objects such as ConfigMaps or Secrets that are meant to be consumed while the build process runs.
+	// After a successful build or when this MachineOSBuild is deleted, these ephemeral objects should be deleted.
```
Should is wooly, be direct, they will be removed, if they aren't, it's a bug
```suggestion
	// After a successful build or when this MachineOSBuild is deleted, these ephemeral objects will be removed.
```

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,208 @@
+	Builder *MachineOSBuilderReference `json:"builder"`
+	// relatedObjects is a list of references to ephemeral objects such as ConfigMaps or Secrets that are meant to be consumed while the build process runs.
+	// After a successful build or when this MachineOSBuild is deleted, these ephemeral objects should be deleted.
+	// However, in the event of a failed build, the objects will not be deleted to allow for inspection and debugging of the failed build process.
```
Is this accurate? Trying to be a bit more direct with the tone
```suggestion
	// In the event of a failed build, the objects will not be remain until the build is removed.
	// This will allow for inspection and debugging of the failed build process.
```

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,208 @@
+	// relatedObjects is a list of references to ephemeral objects such as ConfigMaps or Secrets that are meant to be consumed while the build process runs.
+	// After a successful build or when this MachineOSBuild is deleted, these ephemeral objects should be deleted.
+	// However, in the event of a failed build, the objects will not be deleted to allow for inspection and debugging of the failed build process.
+	// +kubebuilder:validation:MaxItems=10
```
Where does this number come from?Just checking that we are confident in this number

yuqi-zhang replied:
If memory serves, this is based on a discussion with @cheesesashimi .

cheesesashimi replied:
Whenever a build is performed, we currently create multiple ephemeral objects. Right now, it's about 4. I wanted to have some headroom in case more are needed.

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,208 @@
+	// buildStart is the timestamp corresponding to the build controller initiating the build backend for this MachineOSBuild.
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="buildStart is immutable once set"
+	// +optional
+	BuildStart *metav1.Time `json:"buildStart"`
```
You need omitempty on an optional pointer field like this
```suggestion
	BuildStart *metav1.Time `json:"buildStart,omitempty"`
```

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,208 @@
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="buildEnd is immutable once set"
+	// +optional
+	BuildEnd *metav1.Time `json:"buildEnd,omitempty"`
+	// digestedImagePushSpec describes the fully qualified push spec produced by this build.
```
At what point, does an image push spec, become and image pull spec?

yuqi-zhang replied:
Push and pull are relative I suppose. In this case, this is a Push Spec because the image will be pushed to this location upon build completion.

Pull specs are for e.g. the base image, where the build process is pulling them.

JoelSpeed replied:
Feels like a weird nuance, and I think I've asked this several times. IMO, this is likely to cause confusion, but I'm not sure how we can avoid it

JoelSpeed commented on machineconfiguration/v1/tests/machineosbuilds.machineconfiguration.openshift.io/OnClusterBuild.yaml:
```diff
@@ -0,0 +1,275 @@
+            imageBuilderType: Job
+            job:
+              group: batch
+              resource: job
```
Nit, should be `jobs`, not `job`

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,207 @@
+
+// MachineOSConfigSpec describes user-configurable options as well as information about a build process.
+type MachineOSConfigSpec struct {
+	// machineConfigPool is the pool which the build is for
```
What is the MCP used for, gathering information about the rendered machine config? Or? Can we help a user to explain what information we glean by knowing which MCP this is tied to?

yuqi-zhang replied:
This sets the pool which the build will be rolled out to once completed. The build and rollout process are currently coupled, so this is the main definition where the MCO determines what to do with the built image.

JoelSpeed replied:
That context should be added to the godoc

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,207 @@
+	// +kubebuilder:validation:Minimum=0
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// currentImagePullSpec is the fully qualified image pull spec used by the MCO to pull down the new OSImage. This includes the sha256 image digest.
```
Is this updated over time, if the MOSC triggers new builds? Can we explain when this field might change, once populated?

JoelSpeed replied:
This is part of the status of the MOSC, I would expect this to always point to the latest successfully built MOSB, right?

yuqi-zhang replied:
the controller does set it to the MOSB's "push spec": `mosc.Status.CurrentImagePullspec = mosb.Status.FinalImagePushspec`

which I think is populated when a build is successful. So this can be considered tracking the corresponding MOSB with a successfully created push spec

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,207 @@
+	// The length of the whole spec must be between 1 to 447 characters.
+	// +optional
+	CurrentImagePullSpec ImageDigestFormat `json:"currentImagePullSpec,omitempty"`
+	// machineOSBuild is a reference to the MachineOSBuild object for this MachineOSConfig, which contains the status for the image build.
```
Can this change over time? If so, what happens to the old ones?

yuqi-zhang replied:
The mapping should only be 1-to-1, and recreated (rebuilt) if the build object is deleted. cc @cheesesashimi

cheesesashimi replied:
Correct. If the most current MachineOSBuild is deleted, it will be recreated, starting a new build.

JoelSpeed replied:
So, is deleting the build something the controller might do when it wants to trigger a new build? It doesn't leave behind old builds for historical reasons?

yuqi-zhang replied:
Yes, we have an annotation that instructs the controller to delete the current MOSB and kick off a rebuild. The idea being that e.g. a failed build should have been parsed by the user before the kick that off.

JoelSpeed replied:
An annotation on the MOSB? That probably could have been a spec field (we have prior art)

But ok, and, I don't remember now whether this was in the EP or not, but, is there any particular reason why you don't keep history of builds around? If I were a user and wanted to compare several historic builds, with the current behaviour, I could not do that right?

cheesesashimi replied:
> An annotation on the MOSB? That probably could have been a spec field (we have prior art)

We only set it as an annotation because we didn't have a field for it yet. This PR adds a status field (the `MachineOSBuild` object reference field) so we don't need to set this annotation anymore.

> is there any particular reason why you don't keep history of builds around? If I were a user and wanted to compare several historic builds, with the current behaviour, I could not do that right?

It was an oversight. I would like to eventually keep track of history but I don't know what the best practices are around that.

JoelSpeed replied:
> It was an oversight. I would like to eventually keep track of history but I don't know what the best practices are around that.

If that's on the cards, then having a single reference here is probably not what we want. We probably want a slice instead

```
// machineOSBuilds is a reference to the MachineOSBuild objects for this MachineOSConfig.
// This list will contain up to the last 10 builds, with the most recent builds at the top of the list.
MachineOSBuilds []MachineOSBuildHistory `json:"machineOSBuilds,omitempty"`
```

Where the historic item probably needs a little more than the name, roughly perhaps something akin to
```
machineOSBuilds:
- machineOSBuild:
    name: <mosc-name>-<suffix>
  created: <timestamp>
  finished: <timestamp>
  finalImagePushSpec: ...
  buildSummary: Completed | Failed | Pending
```

If we tie ourselves now to a single instance of an object, we can still add this slice later, but you will still always need to populate the current API

JoelSpeed replied:
That's perfectly acceptable as an approach, however, where we've done this before, we have had a list from the beginning, and included the latest build inside the history, see the CVO for example

You have until you ship the GA to decide if you want to change this. In theory, you could tombstone the `machineOSBuild` field between now and 4.19 shipping if you decided to fix the history thing sooner

But once the field is shipped, you must always populate it no matter what additional history you add

yuqi-zhang replied:
Had some discussions around this. I think we're leaning to keep it as two fields:
1. machineOSBuild, which is a reference to the current active build, and
2. machineOSBuildHistory, which is a list of history objects for all builds that have completed.

The idea being that the first field is used to reference the latest active build, and the history is only updated whenever a build completes. For this iteration, we'll just have the former reference for now, and we'll add the history field at a later time once we're ready to implement that.

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,207 @@
+	// +listMapKey=type
+	// +optional
+	// TODO(jerzhang): add godoc after conditions are finalized.
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
```
Conditions should be first in the status

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,207 @@
+	// +listType=map
+	// +listMapKey=type
+	// +optional
+	// TODO(jerzhang): add godoc after conditions are finalized.
```
Check this doesn't end up in generated docs please

yuqi-zhang replied:
I can confirm it doesn't show up in the generated manifests. The generated docs has `TODOs are ignored from the parser` in it as well.

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,207 @@
+	//   RUN rpm-ostree install tree && \
+	//     ostree container commit
+	// +required
+	// +kubebuilder:validation:MaxLength=4096
```
A small sentence explaining the max length is required at the end of the godoc please

JoelSpeed commented:
/test lint

JoelSpeed commented:
/test lint

JoelSpeed commented:
/override ci/prow/lint

The linter failures here are a bug, will get a bug fix up this afternoon for that

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,210 @@
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=machineosbuilds,scope=Cluster,shortName=mosb
```
Shortnames are not allowed, unless you expect the majority of cluster users to frequently access these objects. To my knowledge, this doesn't warrant that.

They create ambiguity in the CLI, since it becomes too easy to overlap. Officially, only fully qualified resource names are supported.

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,210 @@
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=machineosbuilds,scope=Cluster,shortName=mosb
+// +kubebuilder:subresource:status
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/1773
```
```suggestion
// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/2090
```

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,210 @@
+// +openshift:compatibility-gen:level=1
+type MachineOSBuild struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
```suggestion
	
	// metadata is the standard object metadata.
	metav1.ObjectMeta `json:"metadata,omitempty"`
```

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,210 @@
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec describes the configuration of the machine os build
```
```suggestion
	// spec describes the configuration of the machine os build.
	// It is immutable once set.
```

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,210 @@
+// +openshift:compatibility-gen:level=1
+type MachineOSBuildList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
```
```suggestion
	
	// metadata is the standard list metadata.
	metav1.ListMeta `json:"metadata"`
```

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,210 @@
+
+	// items contains a collection of MachineOSBuild resources.
+	// +optional
+	Items []MachineOSBuild `json:"items"`
```
```suggestion
	Items []MachineOSBuild `json:"items,omitempty"`
```

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,211 @@
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=machineosconfigs,scope=Cluster,shortName=mosc
```
No shortnames as explained on the previous type

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,211 @@
+// +kubebuilder:object:root=true
+// +kubebuilder:resource:path=machineosconfigs,scope=Cluster,shortName=mosc
+// +kubebuilder:subresource:status
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/1773
```
```suggestion
// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/2090
```

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
Any printer columns needed for this type?

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,211 @@
+// +openshift:compatibility-gen:level=1
+type MachineOSConfig struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
```suggestion

	// metadata is the standard object metadata.
	metav1.ObjectMeta `json:"metadata,omitempty"`
```

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,211 @@
+// +openshift:compatibility-gen:level=1
+type MachineOSConfigList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
```
```suggestion
	
	// metadata is the standard list metadata.
	metav1.ListMeta `json:"metadata,omitempty"`
```

JoelSpeed commented on machineconfiguration/v1alpha1/types_machineosconfig.go:
```diff
@@ -105,7 +105,7 @@ type BuildInputs struct {
 	// baseImagePullSecret is the secret used to pull the base image.
 	// must live in the openshift-machine-config-operator namespace
-	// +required
+	// +kubebuilder:validation:Required
```
Should reverse that change please

JoelSpeed reviewed (commented):
This is looking very nearly ready to go. Everything I've highlighted now is mechanical and not subjective, so if we can get this all the suggestions merged, then this is good to go

yuqi-zhang commented on operator/v1/tests/machineconfigurations.operator.openshift.io/NodeDisruptionPolicy.yaml:
It's a separate commit added back in since the tests were removed during a cleanup PR at some point it looks like. If you want, I can pull that out of the PR

JoelSpeed commented:
/test integration

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,215 @@
+
+	// status describes the last observed state of this machine os build.
+	// +optional
+	Status *MachineOSBuildStatus `json:"status,omitempty"`
```
No required fields in status, it doesn't need to be a pointer

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,215 @@
+
+	// status describes the status of the machineosconfig
+	// +optional
+	Status *MachineOSConfigStatus `json:"status,omitempty"`
```
No required fields in status, it doesn't need to be a pointer

JoelSpeed commented on machineconfiguration/v1/types_machineosbuild.go:
```diff
@@ -0,0 +1,215 @@
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
Should also be `+optional`

JoelSpeed commented on machineconfiguration/v1/types_machineosconfig.go:
```diff
@@ -0,0 +1,215 @@
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object metadata.
+	metav1.ObjectMeta `json:"metadata,omitempty"`
```
Should also be +optional

JoelSpeed reviewed (commented):
/lgtm

yuqi-zhang commented:
/payload 4.19 ci blocking

JoelSpeed commented:
@yuqi-zhang Are QE happy with the changes as they stand now? Are they able to completely test the feature with this new API and your outstanding MCO PR?

yuqi-zhang commented:
> Are QE happy with the changes as they stand now?

I believe so.

> Are they able to completely test the feature with this new API and your outstanding MCO PR?

Our original plan was to have this and the client-go generation merged, before testing via the graduation/v1 API switch PR and the MCO PR. Currently the MCO PR is only tested via existing unit and layering-e2e tests

JoelSpeed commented:
/lgtm

yuqi-zhang commented:
/payload 4.19 ci blocking

yuqi-zhang commented:
/label acknowledge-critical-fixes-only

Payloads passed
